src/Common/Common.Authentication/WindowsAuthenticationFactory.cs

```diff
@@ -74,14 +75,68 @@ public IAccessToken Authenticate(IPowerBIEnvironment environment, IPowerBILogger
             return token.ToIAccessToken();
         }
 
-        private static TokenCache InitializeCache(IPowerBIEnvironment environment, string queryParams)
+        public IAccessToken Authenticate(IPowerBIEnvironment environment, IPowerBILogger logger, IPowerBISettings settings, string userName, SecureString password)
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                throw new NotSupportedException("Authenticator only works on Windows");
+            }
+
+            LoggerCallbackHandler.UseDefaultLogging = settings.Settings.ShowADALDebugMessages;
+
+            if (Cache == null)
+            {
+                lock (tokenCacheLock)
+                {
+                    if (Cache == null)
+                    {
+                        Cache = InitializeCache(environment, null, userName, password.SecureStringToString());
+                    }
+                }
+            }
+
+            if (!this.AuthenticatedOnce)
+            {
+                throw new AuthenticationException("Failed to authenticate once");
+            }
+
+            var context = new AuthenticationContext(environment.AzureADAuthority, Cache);
+            AuthenticationResult token = null;
+            try
+            {
+                token = context.AcquireTokenSilentAsync(environment.AzureADResource, environment.AzureADClientId).Result;
+            }
+            catch (AdalSilentTokenAcquisitionException)
+            {
+                // ignore and try one more time by getting a new cache and let the exception fly if it fails
+                lock (tokenCacheLock)
+                {
+                    Cache = InitializeCache(environment, null, userName, password.SecureStringToString());
```

Comment:
Consider passing the SecureString directly and calling the SecureStringToString method in InitializeCache.

Reply:
Yeah I will do a refactor